File: recursion/linear_acc/applied/letter_combinations.scala
object Solution {
    def letterCombinations(digits: String): List[String] = {
        // brute force
        // maintain a map for what each letter stands for 
        // iterating over each combination - given that each number represents 3 or 4 letters we have
        // 4 ^ n and 
        val phoneMap = Map(('2' -> "abc"), ('3' -> "def"), ('4' -> "ghi"), ('5' -> "jkl"), ('6' -> "mno"), 
                           ('7' -> "pqrs"), ('8' -> "tuv"), ('9' -> "wxyz"))
        // greedy problem where you can take each element and combine - independently of the other steps
        // can't just recurse from the end because it will be in the wrong order
        def recursiveLetterCombinations(digits: String, acc : List[String]): List[String] = {
            digits match {
                // base case
                case "" => acc 
                // recursive case
                case digits => 
                    recursiveLetterCombinations(digits.tail, 
                                               // ["ad", "ae"...etc] -> "ad" + "a" or  "b" or "c"
                                               acc.flatMap(ans => phoneMap(digits.head).map(elem => ans + elem)))
            }
        }
        if(digits == "") 
            List()
        else
            recursiveLetterCombinations(digits, List(""))
        
    }
}

File: data_structures/recursion/divide_and_conquer_accumulate/combinations/backtrack/letter_combinations.go
package data_structures

func letterCombinations(digits string) []string {
	// base case
	if len(digits) == 0 {
		return []string{}
	}

	digitsMap := map[string]string{
		"2": "abc",
		"3": "def",
		"4": "ghi",
		"5": "jkl",
		"6": "mno",
		"7": "pqrs",
		"8": "tuv",
		"9": "wxyz",
	}

	d := data{
		digits:       digits,
		digitsMap:    digitsMap,
		combinations: []string{},
	}
	d.backtrack(0, "")
	return d.combinations
}

type data struct {
	digits       string
	digitsMap    map[string]string
	combinations []string
}

func (d *data) backtrack(i int, path string) {
	// base case
	if len(path) == len(d.digits) {
		d.combinations = append(d.combinations, path)
		return
	}
	// recursive case
	digit := d.digits[i]
	for _, char := range d.digitsMap[string(digit)] {
		path += string(char)
		d.backtrack(i+1, path)
		path = path[:len(path)-1]
	}
}
